### PR TITLE
[SPARK-8999][MLlib]Support non-temporal sequence in PrefixSpan (Array[Array[Int]])

### PR DESCRIPTION
**Pull request by zhangjiajin**
Support non-temporal sequence in PrefixSpan

### CODE REVIEW

**Comment by feynmanliang** [comment]
jenkins retest this please

**Comment by mengxr** [comment]
@zhangjiajin I'm a little worried about the performance/scalability if we use `Array[Array[Int]]` to represent a sequence. It might create too many objects when the dataset is large. An alternative is keeping `Array[Int]` and use `-1` to separate itemsets, which is used by SPMF. If we support generic item type, we can use `null` to separate itemsets. Could you test this approach?

**Comment by zhangjiajin** [comment]
@mengxr OK. I will test two approaches (Array[Array[Int]] and Array[Int]).

**Comment by zhangjiajin** [comment]
@mengxr I compared two approaches (Array[Array[Int]] and Array[Int]), I found Array[Array[Int]] performance is better than Array[Int].
The dataset I used is BMSWebView2( KDD CUP 2000). When the support is 1500, the running time of Array[Array[Int]] is 44s, the running time of Array[Int] is 69s. These time values are the average of three measurements.

File: mllib/src/test/scala/org/apache/spark/mllib/fpm/PrefixSpanSuite.scala [hunk 1]
```diff
@@ -42,13 +42,13 @@ class PrefixSpanSuite extends SparkFunSuite with MLlibTestSparkContext {
       Array(3, 4, 4, 3),
       Array(6, 5, 3))
 
-    val rdd = sc.parallelize(sequences, 2).cache()
+    val rdd = sc.parallelize(sequences.map(_.map(Array(_))), 2).cache()
 
     def compareResult(
```
Comment (feynmanliang) [inline mllib/src/test/scala/org/apache/spark/mllib/fpm/PrefixSpanSuite.scala]: Move helper method outside of test

File: mllib/src/test/scala/org/apache/spark/mllib/fpm/PrefixSpanSuite.scala [hunk 2]
```diff
@@ -107,6 +107,75 @@ class PrefixSpanSuite extends SparkFunSuite with MLlibTestSparkContext {
...
-    assert(compareResult(expectedValue3, result3.collect()))
+    assert(compareResult(expectedValue3.map(x => (x._1.map(Array(_)), x._2)), result3.collect()))
+
+    val sequences4 = Array(
+      "a,abc,ac,d,cf",
+      "ad,c,bc,ae",
+      "ef,ab,df,c,b",
+      "e,g,af,c,b,c")
+    val coder = Array('a', 'b', 'c', 'd', 'e', 'f', 'g').zip(Array(1, 2, 3, 4, 5, 6, 7)).toMap
+    val intSequences = sequences4.map(_.split(",").map(_.toArray.map(coder)))
```
Comment (feynmanliang) [inline mllib/src/test/scala/org/apache/spark/mllib/fpm/PrefixSpanSuite.scala]: Please remove `coder` and test only `PrefixSpan`'s implementation in unit tests. We have a `StringIndexer` to do what `coder` is doing here.

File: mllib/src/test/scala/org/apache/spark/mllib/fpm/PrefixSpanSuite.scala [hunk 3]
```diff
@@ -42,13 +42,13 @@ class PrefixSpanSuite extends SparkFunSuite with MLlibTestSparkContext {
       Array(3, 4, 4, 3),
       Array(6, 5, 3))
 
-    val rdd = sc.parallelize(sequences, 2).cache()
+    val rdd = sc.parallelize(sequences.map(_.map(Array(_))), 2).cache()
```
Comment (feynmanliang) [inline mllib/src/test/scala/org/apache/spark/mllib/fpm/PrefixSpanSuite.scala]: Why do we need the `.map`s?

File: mllib/src/test/scala/org/apache/spark/mllib/fpm/PrefixSpanSuite.scala [hunk 4]
```diff
@@ -107,6 +107,75 @@ class PrefixSpanSuite extends SparkFunSuite with MLlibTestSparkContext {
       (Array(4, 5), 2L),
       (Array(5), 3L)
     )
-    assert(compareResult(expectedValue3, result3.collect()))
+    assert(compareResult(expectedValue3.map(x => (x._1.map(Array(_)), x._2)), result3.collect()))
+
+    val sequences4 = Array(
```
Comment (feynmanliang) [inline mllib/src/test/scala/org/apache/spark/mllib/fpm/PrefixSpanSuite.scala]: Is this a test for non-temporal sequences? If so, please move to a separate `test` block

**Pull request closed without merging**